### PR TITLE
Adding base implementation of an EditorHover listener

### PR DESCRIPTION
**Pull request by jyoo980**
* Mainly extracting functionality already implemented in my fork of IntelliJ
    CE.
  * Able to extract the PsiElement under the hover.
  * Added extension methods to `PsiElement` that determine whether an element is a:
     * Non-literal method argument.
     * Non-literal local variable

### CODE REVIEW

File: src/main/kotlin/com/github/jyoo980/reachhover/listeners/EditorHoverListener.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import com.intellij.openapi.editor.event.EditorMouseEvent
+import com.intellij.openapi.editor.event.EditorMouseMotionListener
+
+internal class EditorHoverListener : EditorMouseMotionListener {
+
+    private val logger: Logger = Logger.getInstance(EditorHoverListener::class.java)
+
+    override fun mouseMoved(e: EditorMouseEvent) {
+        val offset = MouseHoverUtil.targetOffset(e) ?: return
+        val project = e.editor.project ?: return
```
Comment (jyoo980) [inline src/main/kotlin/com/github/jyoo980/reachhover/listeners/EditorHoverListener.kt:14]: I could do something like 
```
MouseHoverUtil.targetOffset(e)?.let { offset ->
  e.editor.project?.let { project ->
    ...
}
```
But I feel like that'll get pretty messy if we end up having more than 2 nullables.

I know in Scala you could have for-comprehensions, like

```scala
for {
  a <- maybeA
  b <- maybeB
} yield ...
```
But I'm not sure if there's anything like that in Kotlin.

File: src/main/kotlin/com/github/jyoo980/reachhover/util/PsiExtension.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * Extension method on PsiElement.
+ *
+ * @return true iff the element is an instance of a local variable reference.
+ */
+fun PsiElement.isLocalVariableReference(): Boolean {
+    val isParentLocalVar =
+        this.parent?.let { PsiTreeUtil.instanceOf(it, PsiLocalVariable::class.java) } ?: false
+    return isParentLocalVar && PsiTreeUtil.instanceOf(this, PsiIdentifier::class.java)
```
Comment (SmirnovOleg) [inline src/main/kotlin/com/github/jyoo980/reachhover/util/PsiExtension.kt:31]: This extension could look like this, couldn't it?
```
fun PsiElement.isLocalVariableReference(): Boolean {
    return this is PsiIdentifier && parent is PsiLocalVariable
}
```

File: src/main/kotlin/com/github/jyoo980/reachhover/util/PsiExtension.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+fun PsiElement.isNonLiteralMethodArg(): Boolean {
+    val optParentMethodCallExpr =
+        PsiTreeUtil.getParentOfType(this, PsiMethodCallExpression::class.java)
+    // Need to check if one of the element's parents is a method list. Otherwise, we get items like
+    // a.foo() being erroneously reported.
+    val optParentExprList = PsiTreeUtil.getParentOfType(this, PsiExpressionList::class.java)
+    return PsiTreeUtil.instanceOf(this, PsiIdentifier::class.java) &&
+        optParentMethodCallExpr != null &&
+        optParentExprList != null
```
Comment (SmirnovOleg) [inline src/main/kotlin/com/github/jyoo980/reachhover/util/PsiExtension.kt:20]: I'd rather prefer something like this, but it's more a matter of taste.
```
    PsiTreeUtil.getParentOfType(this, PsiMethodCallExpression::class.java) ?: return false
    PsiTreeUtil.getParentOfType(this, PsiExpressionList::class.java) ?: return false
    return PsiTreeUtil.instanceOf(this, PsiIdentifier::class.java)
```

File: src/main/kotlin/com/github/jyoo980/reachhover/util/MouseHoverUtil.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    fun elementAtOffset(project: Project, editor: Editor, offset: Int): PsiElement? {
+        val optFile = PsiDocumentManagerImpl.getInstance(project).getPsiFile(editor.document)
+        return optFile?.let { it ->
+            var element = it.findElementAt(offset)
+            if (element == null && offset == it.textLength) {
+                element = it.findElementAt(offset - 1)
+            }
+            if (element is PsiWhiteSpace) {
+                element = it.findElementAt(element.textRange.startOffset - 1)
```
Comment (SmirnovOleg) [inline src/main/kotlin/com/github/jyoo980/reachhover/util/MouseHoverUtil.kt:41]: `element.prevSibling`?

File: src/main/kotlin/com/github/jyoo980/reachhover/listeners/EditorHoverListener.kt (hunk 1 above)
Comment (SmirnovOleg) [reply src/main/kotlin/com/github/jyoo980/reachhover/listeners/EditorHoverListener.kt:14]: I don't know such a way as well. But it doesn't mean that it doesn't exist :) 
I think the way with `?: return` is just fine.

**Review by SmirnovOleg — APPROVED** [review]

File: src/main/kotlin/com/github/jyoo980/reachhover/util/PsiExtension.kt (hunk 3 above)
Comment (jyoo980) [reply src/main/kotlin/com/github/jyoo980/reachhover/util/PsiExtension.kt:20]: Definitely seems more idiomatic, I like it

**Pull request merged**